pkg/test/parameter.go

```diff
@@ -7,23 +7,46 @@ package test
 
 import (
 	"bytes"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"text/template"
 
 	"github.com/facebookincubator/contest/pkg/target"
 )
 
-// Param represents a test step parameter. It is initialized from a string,
-// which can be either a host name, a function expression, or just any regular
-// string. It enables hostname and function expression validation and a few
-// other validation operations. This is meant to simplify the way plugin access
-// and use test step parameters.
-type Param string
+// NewParam inititalizes a new Param object from a string
+func NewParam(s string) Param {
+	var p Param
+	p.RawMessage = json.RawMessage(s)
```

Comment:
This takes random string content and pretends it is valid JSON.
This only works because `Param.String()` below returns the raw message as string if `Unmarshal` fails.

Would it be a better to:
- Check if the input is valid json, then accept
- Otherwise tread as random string and marshal into json. This can fail. Not sure what to do then.

Reply:
Shouldn't `Param` expose both `AsString` (with `string, error` signature) which fails if we cannot unmarshal into string, and `AsJSON` (with `json.RawMessage, error` signature) which instead fails if we can marshal into string? 

 The plugin should have clear expectations on what the parameters should look like. It might event decide to accept any of the two formats, then try both `AsString` and `AsJSON`, and be given a clear signal by `Param` based on the return values on which one is valid, and act accordingly.

Reply:
We can do that, but it makes calling AsString() very annoying for the simple usecase, so it clutters the plugin code. I think @insomniacslk wanted to prevent that. The current API with `GetOne` goes into a similar direction: It never fails, but might return empty strings.
I am not married to this, I can totally change it. What does @insomniacslk think?

Reply:
I personally think `AsString` should just return `string(p.RawMessage)`, without trying to interpret it (at the end of the day, the JSON subdocument is a string, and it's again up to the plugin to interpret it if it's not a plain string).

What would `AsJSON` take as input, and return? I'm not sure I understand what its role is
CC @marcoguerri

> I think @insomniacslk wanted to prevent that.

In general, yes, if I can choose I prefer to move the complexity away from users and plugin developers, but of at the end of the day we have to do what makes sense, if there's no better way it's OK to revert the above principle

Reply:
Decided to keep it simple for now, as this should only be relevant for unit tests